packages/buidler-core/src/internal/buidler-evm/provider/filter.ts

```diff
@@ -0,0 +1,130 @@
+import Bloom from "@nomiclabs/ethereumjs-vm/dist/bloom";
+import { BN, bufferToHex, toBuffer } from "ethereumjs-util";
+
+import { RpcLogOutput } from "./output";
+
+export const LATEST_BLOCK = -1;
+
+export enum Type {
+  LOGS_SUBSCRIPTION = 0,
+  PENDING_TRANSACTION_SUBSCRIPTION = 1,
+  BLOCK_SUBSCRIPTION = 2
+}
+
+export interface FilterCriteria {
+  fromBlock: number;
+  toBlock: number;
+  addresses: Buffer[];
+  topics: Array<Array<Buffer | null> | null>;
+}
+
+export interface Filter {
+  id: number;
+  type: Type;
+  criteria?: FilterCriteria;
+  deadline: Date;
+  hashes: string[];
+  logs: RpcLogOutput[];
+  subscription: boolean;
+}
+
+export function bloomFilter(
+  bloom: Bloom,
+  addresses: Buffer[],
+  topics: Array<Array<Buffer | null> | null>
+): boolean {
+  if (addresses.length > 0 && !bloom.multiCheck(addresses)) {
+    return false;
+  }
+
+  for (const sub of topics) {
+    if (sub == null) {
+      continue;
+    }
+
+    let included = sub.length === 1;
```

Comment:
I don't get this part. What if we have a single topic that isn't included (`bloom.check(t) === false`)? Wouldn't `included` be `true`?

Reply:
Oh, I guess that should have been a `0`. 

Maybe something like this would be simpler:

```ts
    if (sub.length === 0) {
      continue;
    }

    let included = false;
    for (const topic of sub) {
      if (topic != null && bloom.check(topic)) {
        included = true;
        break;
      }
    }
```

Reply:
👍